Multiple entities on history panel

D3v01dZA opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
Allow multiple entities to be selected on the history panel. I have found that for some use cases I want to see all entities together on a timeline and couldn't find any way to do it. On investigation of the history API I discovered that it is already possible to query the state of multiple entities using a comma-separated list. I switched out the entity-picker with an entities-picker and enhanced entities-picker to be displayed horizontally so that it looks somewhat neat.

<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue or discussion
  in the additional information section.
-->

## Type of change
User interface only modification to allow the user to select multiple entities on the history page instead of just one.

<!--
  What type of change does your PR introduce to the Home Assistant frontend?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information

<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- I didn't create an issue or anything to initially gather thoughts so let me know if I need to do something like that!

## Checklist

<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

<!--
  Thank you for contributing <3
-->

[docs-repository]: https://github.com/home-assistant/home-assistant.io

bramkragten commented:
Can you provide a screenshot of the new screen?

D3v01dZA commented:
Thanks for the quick reply! Here are a couple using the browser and some on an emulated mobile device.

![Browser Entities - None](https://user-images.githubusercontent.com/10627822/132044013-f49adbf6-6635-4adc-8e37-968254862474.png)
![Browser Entites- One](https://user-images.githubusercontent.com/10627822/132044011-a6594b5f-5190-4b04-9930-7d815f66ba7f.png)
![Browser Entities - Multiple](https://user-images.githubusercontent.com/10627822/132044012-5b09d55b-17f7-4c3d-9375-d26a3112cf09.png)
![Browser Entites - Lots](https://user-images.githubusercontent.com/10627822/132044009-b324bd6a-4fb2-492c-9998-ec7b3234ef44.png)
![Mobile Entites - None](https://user-images.githubusercontent.com/10627822/132044014-ae5a5d79-2c7d-45fa-9af7-8e11fc48ff2f.png)
![Mobile Entites - One](https://user-images.githubusercontent.com/10627822/132044016-f9cbccd2-2b90-4ddc-9690-b1c63b252eb4.png)
![Mobile Entities - Multiple](https://user-images.githubusercontent.com/10627822/132044017-5c331508-a335-4fde-ab47-138f17da98f1.png)

bramkragten commented:
Not too happy with the UI here, a checkbox/checked list inside the dropdown would probably work a lot better. Adding our UX designer @matthiasdebaat here for a little help (he is on vacation atm, so might take some time for him to respond)

D3v01dZA commented:
Thanks for the feedback. That was my initial idea too, this implementation also doesn't handle re-selecting the same entity in a different box very well (but I have this gripe with everywhere that the multiple-entity select is used).

I did this initially to get the ball rolling by re-using existing component, and timing is not really an issue, I actually think I may take the time to see if I can do this in a single box instead, but to do that I don't think I'll be re-using either the current multiple-entity or single-entity selects.

homeassistant commented:
Hello @D3v01dZA,

When attempting to inspect the commits of your pull request for CLA signature status among all authors we encountered commit(s) which were not linked to a GitHub account, thus not allowing us to determine their status(es).

The commits that are missing a linked GitHub account are the following:

- [`aff377acee99c0841a72765a4f5a0bf3016cc75e`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;frontend&#x2F;pull&#x2F;9946&#x2F;commits&#x2F;aff377acee99c0841a72765a4f5a0bf3016cc75e) - This commit has something that looks like an email address (me+git@caltona.net). Maybe try linking that to GitHub?.

Unfortunately, **we are unable to accept this pull request until this situation is corrected.**

Here are your options:

1. If you had an email address set for the commit that simply wasn't linked to your GitHub account you can link that email now and it will retroactively apply to your commits. The simplest way to do this is to click the link to one of the above commits and look for a blue question mark in a blue circle in the top left. Hovering over that bubble will show you what email address you used. Clicking on that button will take you to your email address settings on GitHub. Just add the email address on that page and you're all set. GitHub has more information about this option [in their help center](https://help.github.com/articles/setting-your-email-in-git/#commits-on-github-arent-linking-to-my-account).

2. If you didn't use an email address at all, it was an invalid email, or it's one you can't link to your GitHub, you will need to change the authorship information of the commit and your global Git settings so this doesn't happen again going forward. GitHub provides some great instructions on how to change your authorship information [in their help center](https://help.github.com/articles/setting-your-email-in-git/).
   * If you only made a single commit you should be able to run
      ```
      git commit --amend --author="Author Name <email@address.com>"
      ```
      (substituting Author Name and email@address.com for your actual information) to set the authorship information.
   * If you made more than one commit and the commit with the missing authorship information is not the most recent one you have two options:
       1. You can re-create all commits missing authorship information. This is going to be the easiest solution for developers that aren't extremely confident in their Git and command line skills.
       2. You can use [this script](https://help.github.com/articles/changing-author-info/) that GitHub provides to rewrite history. **Please note:** this should be used only if you are very confident in your abilities and understand its impacts.
   * Whichever method you choose, I will come by to re-check the pull request once you push the fixes to this branch.

We apologize for this inconvenience, especially since it usually bites new contributors to Home Assistant. We hope you understand the need for us to protect ourselves and the great community we all have built legally. The best thing to come out of this is that you only need to fix this once and it benefits the entire Home Assistant and GitHub community.

Thanks, I look forward to checking this PR again soon! :heart:

D3v01dZA commented:
I toyed with this again today, I have it working with a single combo-box, so now you would select all of the entites you want from a single list. It's not working 100% as I am really not familiar with vaadin and I assume that like other select's its a bit of a mission to make it display something which is not a single selectable entity (which would be all the names of the selected.

I did not take this any further than showing it could work though as I would rather wait for the UX designer than spend time doing something they would not want.

![Single Box](https://user-images.githubusercontent.com/10627822/132140006-56d2962a-79dd-44fc-a889-a27ec4b6cd2b.png)

spacegaier commented:
Two comments:
1. Rather than just selecting entities, wouldn't it make more sense to use the same picker that we e.g. have in the automation editor that allows also selecting via device and area? That would also fulfill this FR https://github.com/home-assistant/frontend/discussions/7825 plus already supports selecting many entities out of the box
![image](https://user-images.githubusercontent.com/114137/132144677-10b0d337-26f8-4691-8433-57deda19bba7.png)

2. Whatever we end up doing for the history panel should then also be added to the logbook panel for consistency.

D3v01dZA commented:
I actually really think I like using this picker instead @spacegaier , I must confess that I haven't used these pickers much (mostly directly use YAML) which is why it slipped my mind!

matthiasdebaat commented:
Great idea to filter multiple entities @D3v01dZA 

> Rather than just selecting entities, wouldn't it make more sense to use the same picker that we e.g. have in the automation editor that allows also selecting via device and area?

Good suggestion. I've made a small mockup of the current, `chips` and combobox selector. https://www.figma.com/file/CkkpnnyN1T5zI8W4VTsX5O/History-Logbook-entity-filter?node-id=0%3A1

In the automation editor, the input appears above the 'chips' and pushes the content down. For the History and Logbook page this isn't very nice. On desktop we were able to show the input to the left of the `chips`. On mobile, there's no choice but to move the content.

Personally, I like the combo box better, as there is only one chip and this interaction pattern makes it more difficult to select multiple entities.

spacegaier commented:
In my mind I would have added the complete chip section as we have in the automation editor, so including the device and the area selector since I think that can also be helpful for history and logbook.

matthiasdebaat commented:
> In my mind I would have added the complete chip section as we have in the automation editor, so including the device and the area selector since I think that can also be helpful for history and logbook.

I thought there are only "entities" in history? Thats why I removed `areas` and `devices`. Or am I wrong?

spacegaier commented:
> 
> 
> > In my mind I would have added the complete chip section as we have in the automation editor, so including the device and the area selector since I think that can also be helpful for history and logbook.
> 
> I thought there are only "entities" in history? Thats why I removed `areas` and `devices`. Or am I wrong?

That's true, but in my mind selecting an area would then show all histories of entities belonging to that area. Same for devices. That is what the selectors in the service editor do as well in the backend.

D3v01dZA commented:
I've been toying with this in my head a bit and there are two things. First I think the combobox has the upper hand when it comes to layout, it's much simpler to have one box with a growing list of selected entities than a physically growing list of entities that potentially overflows at a certain point (although you can make that look fairly decent). Secondly it might be nice to have some sort of domain or measurement type chip, for example I was trying to investigate how my AC cools the whole house so I would have said show me temperatures in x, y, and z areas, or even just temperatures in general so I can contrast the various sensors I have all at once.

bramkragten commented:
> > > In my mind I would have added the complete chip section as we have in the automation editor, so including the device and the area selector since I think that can also be helpful for history and logbook.
> > 
> > 
> > I thought there are only "entities" in history? Thats why I removed `areas` and `devices`. Or am I wrong?
> 
> That's true, but in my mind selecting an area would then show all histories of entities belonging to that area. Same for devices. That is what the selectors in the service editor do as well in the backend.

Yeah, we have an expand function for devices and areas to get the entities, so that would be possible

matthiasdebaat commented:
> > > In my mind I would have added the complete chip section as we have in the automation editor, so including the device and the area selector since I think that can also be helpful for history and logbook.
> > 
> > 
> > I thought there are only "entities" in history? Thats why I removed `areas` and `devices`. Or am I wrong?
> 
> That's true, but in my mind selecting an area would then show all histories of entities belonging to that area. Same for devices. That is what the selectors in the service editor do as well in the backend.

Yeah that's nice! Let's go with this.

D3v01dZA committed:
Use the chips selector on the history panel

Place the chips below the dates always

Layout the history panel chips horizontally

Update src/components/chart/state-history-charts.ts

Thanks!

Co-authored-by: Zack Barett <zackbarett@hey.com>

Add support for query params on the picker

D3v01dZA commented on src/panels/history/ha-panel-history.ts:
```diff
@@ -174,6 +187,48 @@ class HaPanelHistory extends LitElement {
     this._isLoading = false;
   }
 
+  private _getEntityIds(): string[] {
```
I'm sure there's a way better way to do this, but I only use JS/TS for my toy projects so I really don't know how to handle this any better, advice is appreciated!

D3v01dZA commented:
So I added changes to use the chips selector. I did have to write my own method to extract the entity ids, it seems that @spacegaier may be correct but the service page seems to do the translation from device/area to entity in the backend as it sends the area and device as is.

There are definitely some issues with layout, as you can see in the screenshots when you select something, using vertical space looks strange next to the combobox. I think I should put the selected items to the right of the chips rather than above to solve this but I will wait for more @matthiasdebaat to suggest something before I do anything else. You will also note that when it overflows it moves the whole box.

![Chips - No Selection](https://user-images.githubusercontent.com/10627822/134778157-4c05d8af-60b3-4deb-9e42-b27b484f7592.png)
![Chips - Area](https://user-images.githubusercontent.com/10627822/134778156-d2b8b601-8dbc-4c96-85df-945df648626b.png)
![Chips - Overflow](https://user-images.githubusercontent.com/10627822/134778158-75175280-573e-4069-9e1f-0aa85709966a.png)

matthiasdebaat commented:
> There are definitely some issues with layout, as you can see in the screenshots when you select something, using vertical space looks strange next to the combobox. I think I should put the selected items to the right of the chips rather than above to solve this but I will wait for more @matthiasdebaat to suggest something before I do anything else. You will also note that when it overflows it moves the whole box.

Nice, I like what you did here! I agree, let's move the selected items to the right. What will happen if there are a lot of selections?

D3v01dZA commented:
If you have a look at the last screenshot, it moves the whole chip-selector to the next line, its quite jarring :)

D3v01dZA commented:
So I tried to move them to the side, but it doesn't work all that well, each chip still pulls up a combobox above it vertically and this still messes with verticality. Placing everything to the side also doesn't work because it still jumps, only now it does so to the right. So what I will suggest is to place the chips selector directly below the dates always (and I added a bit of padding on large screens to try to align a bit nicer).

Let me know if there is a better way to do this @matthiasdebaat, I am not entirely happy with this (but its not particularly bad) but I'd need a concrete location on where to put the date selector, chip selectors, combo-box (and combo-box when items are selected) and the actual selected items to decide how to make another change.

I also fixed a small bug where the history line wasn't respecting noSingle from its parent.

![Picker Below](https://user-images.githubusercontent.com/10627822/135354911-79fbdd42-6e23-449d-8f17-422283fa16bd.png)

matthiasdebaat commented:
Could you make it like this? A next row for the selected items when it doesn't fit on one?

![History + Chips](https://user-images.githubusercontent.com/20203284/135423652-3a54503a-8b67-464e-a379-7417960b4c4c.png)

When the combobox is shown it could work like this.
![History + Chips with input](https://user-images.githubusercontent.com/20203284/135424651-02d236b6-543b-445c-92ce-049831e58cbe.png)

bramkragten commented:
Maybe it is nice to do the combobox in a dialog/overlay then?

D3v01dZA commented:
So I changed the layout to be horizontal and I prevented the jumping by hiding the combo-box entirely. It is displayed using absolute with a height of 0px so will always be right under the chips when you click on one of them. The selected entities are then populated to the right of the chips. There is still one jump when you overflow a line, but its handled more or less gracefully I think and its impossible to not jump at all when there is simply too much data. Let me know what you think! (I like the removal of the select in general tbh, I think it looks a bit better than the current select that pops into existence)

![Picker Open](https://user-images.githubusercontent.com/10627822/136704321-419ecf44-5c87-4f95-a932-a6f4010a5ded.png)
![Selections Made](https://user-images.githubusercontent.com/10627822/136704323-fb11a505-0e6e-40dc-8b35-ecdcd183ffed.png)

zsarnett commented:
@D3v01dZA Sorry this has been waiting around for so long. Big change so its hard to review.

I'll take a look at this but would you be able to resolve the merge conflicts?

D3v01dZA commented:
That's quite OK @zsarnett I'll try to have a look over the weekend to get it through. I understand that there are an insane amount of changes going through HA in general!

D3v01dZA commented:
@zsarnett it took me quite a bit longer than I expected to get to this. I updated it, it looks like query params were added to the page to specify an entity id, so I allowed the picker to accept that but did not add the extra params like area and entity, which I could likely do if its desired

zsarnett commented:
😬  Well we are playing ping pong here. Sorry about that. There are more conflicts that need to be addressed

zsarnett commented:
I also think that on the initial load we should not show any history. But that can be used in a different PR

D3v01dZA commented:
It does appear that way - I think I can quite easily prevent the load at initial page opening, I'll update this as I get time

D3v01dZA commented on src/data/history.ts:
```diff
@@ -185,16 +185,12 @@ export const fetchDate = (
     `history/period/${startTime.toISOString()}?end_time=${endTime.toISOString()}&minimal_response${
-      entityId ? `&filter_entity_id=${entityId}` : ``
-    }${
-      entityId && !entityIdHistoryNeedsAttributes(hass, entityId)
```
Realistically its quite hard to determine if we need to fetch attributes or not when querying for a collection. At most I think I _could_ run through all of the entity ids and if none of them need attributes not retrieve the attributes, but I am actually not too certain, so any guidance is appreciated.

bdraco replied:
You'll need to set the flag to true if any entity needs it in the query since it will tell the backend to not pull the data in the query as its not smart enough to union two queries together for ones that need attributes and one that don't so if any entity needs it you can't pass no_attributes.

Its possible to optimize it but we have so much branching logic on the backend to handle old database formats its not something I want to do right now

D3v01dZA replied:
I'm not fully certain I understand, are you saying that right now it won't work correctly because I haven't set the no_attributes flag, or that it will be an inefficient call when no attributes are required?

bdraco replied:
It will be inefficient if you don't set no_attributes because it has to select a lot of data from the database, but you can't set it if any of the selected entities require attributes

D3v01dZA commented:
@zsarnett rebased, with one merge conflict that I wasn't sure of resolving. It works as is but I think that the part I left out is mostly an optimization, but your confirmation would help me out. I also stopped the history panel from loading everything on first load, so if there is nothing selected and nothing comes in on the URL, nothing will be displayed.

fjufju commented:
How I can test this local? This look ready and would be nice to have this in next release.
@zsarnett how can I help you with testing so this can be accepted.

bdraco commented:
Also the logbook backend now supports selecting an arbitrary number of entities and devices

D3v01dZA commented:
@bdraco I'm assuming you're suggesting doing the same thing for logbook?

bdraco commented:
> @bdraco I'm assuming you're suggesting doing the same thing for logbook?

At some point yes

D3v01dZA commented:
I realized that some entities were not populating in some cases, it turns out that the subscribe function doesn't return every possible entity and some things called "state entities" also exist, so I had a look around to find how to discover them and implemented that too. An example of this entity I didn't know existed is sun.sun.

@zsarnett rebased again, hopefully we can get someone to review this so we can get it pushed through. I'm not 100% clear on how to proceed from here but I'd like to see this merged as I think it is a valuable change! (Certainly I could use it)

@bdraco I updated the PR to only fetch attributes if at least one entity requires it, please let me know if I got it right or not!

@fjufju I'm not 100% sure how to deploy this on a up to date system, but my best guess would be to follow this [Dev Docs](https://developers.home-assistant.io/docs/frontend/development). So pull my branch into a repo, build it, and then tell your actual Home Assistant where to get it from, just like is done for development.

D3v01dZA committed:
Remove some code duplication

zsarnett commented:
When no entities are selected, can we show a hint to what they need to do instead?

"Select what you would like to see history for above"

Or something to that matter

zsarnett commented on src/panels/history/ha-panel-history.ts:
```diff
@@ -255,6 +371,18 @@ class HaPanelHistory extends LitElement {
+        :host(:not([narrow])) .selector-padding {
+          padding-left: 32px;
+        }
+
```
for the `.filters` class.

Lets make that `align-items: flex-start`

zsarnett commented on src/panels/history/ha-panel-history.ts:
```diff
@@ -80,25 +100,40 @@ class HaPanelHistory extends LitElement {
+              .hass=${this.hass}
+              .value=${this._targetPickerValue}
+              .disabled=${this._isLoading}
+              .horizontal=${true}
```
```suggestion
              horizontal
```

D3v01dZA committed:
Update src/panels/history/ha-panel-history.ts

Co-authored-by: Zack Barett <zackbarett@hey.com>

bdraco commented on src/data/history.ts:
```diff
@@ -223,36 +223,32 @@ export const fetchDate = (
   hass: HomeAssistant,
```
Is `fetchDate` dead code now?

D3v01dZA replied:
I think it may be, but I didn't remove it, it was changed to fetchDateWS through the duration of this PR

D3v01dZA replied:
By a concurrent commit

bdraco commented:
<img width="1275" alt="Screen Shot 2022-06-29 at 08 51 12" src="https://user-images.githubusercontent.com/663432/176453456-d244f1ee-bf7f-4243-8a2e-204e3f3f5271.png">

Tested and it seems to all work as expected.  I'm not sure if we can do anything about the dead space above the data picker

zsarnett commented:
One of my comments above addresses that. I will most likely merge this and make the changes if the PR Owner doesn't before beta

zsarnett commented:
I want to get this into Beta. Going to merge we can do a new PR to fix the comments as they aren't blocking or broken

bdraco commented:
Finished testing this.  Entities and Devices working as expected.  Areas do not work though

mbo18 commented:
Please add a "See all" button or as in previous release, by default, just display all. This panel is useless now and selecting 10 entities (like 10 temperature sensors) from different areas is just way to complicated
